modules/siddhi-core/src/main/java/org/wso2/siddhi/core/query/selector/attribute/aggregator/UnionAttributeAggregator.java

```diff
@@ -0,0 +1,170 @@
+                        "stock symbols received during a sliding window of 10 seconds."
+        )
+)
+public class UnionAttributeAggregator extends AttributeAggregator {
+
+    /**
+     * This map aggregates the count per each distinct element
+     */
+    private Map<Object,Integer> counter = new HashMap<>();
+    private Set set = new HashSet();
+
+    /**
+     * The initialization method for FunctionExecutor
+     *
+     * @param attributeExpressionExecutors are the executors of each attributes in the function
+     * @param configReader                 this hold the {@link CountAttributeAggregator} configuration reader.
+     * @param siddhiAppContext             Siddhi app runtime context
+     */
+    @Override
+    protected void init(ExpressionExecutor[] attributeExpressionExecutors, ConfigReader configReader,
+                        SiddhiAppContext siddhiAppContext) {
+        if (attributeExpressionExecutors.length != 1) {
+            throw new OperationNotSupportedException("union aggregator has to have exactly 1 parameter, currently " +
+                    attributeExpressionExecutors.length + " parameters provided");
+        }
+        if (attributeExpressionExecutors[0].getReturnType() != Attribute.Type.OBJECT) {
+            throw new OperationNotSupportedException("Parameter passed to union aggregator should be of type" +
+                    " object but found: " + attributeExpressionExecutors[0].getReturnType());
+        }
+    }
+
+    public Attribute.Type getReturnType() {
+        return Attribute.Type.OBJECT;
+    }
+
+    @Override
+    public Object processAdd(Object data) {
+        Set inputSet = (Set) data;
+        for (Object o : inputSet) {
+            set.add(o);
+            if (counter.get(o) == null) {
+                counter.put(o, 1);
+            } else {
+                counter.put(o, counter.get(o) + 1);
+            }
+        }
+        // Creating a new set object as the returned set reference is kept until the aggregated values are
+        // inserted into the store
+        Set returnSet = new HashSet();
+        returnSet.addAll(set);
+        return returnSet;
+    }
+
+    @Override
+    public Object processAdd(Object[] data) {
+        //Union can have only one input parameter, hence this will not be invoked.
+        return null;
+    }
+
+    @Override
+    public Object processRemove(Object data) {
+        Set newSet = (Set) data;
+        for (Object o : newSet) {
+            Integer currentCount = counter.get(o);
+            if (currentCount == null) {
+                //means o does not exist in the counter map or in the set hence doing nothing
+            } else if (currentCount == 0) {
+                throw new IllegalStateException("Error occurred when removing element from " +
+                        "union-set for element: " + o.toString());
+            } else  {
+                counter.put(o, counter.get(o) - 1);
```

Comment:
Optimize the code, don't do "counter.get(o)" always

check if (currentCount == 1) 
then remove the object from set 
else do
counter.put(o, currentCount - 1);

Reply:
Done in 228a774